interfaces/builtin/usb_gadget.go

```diff
@@ -29,25 +48,199 @@ const usbGadgetBaseDeclarationSlots = `
+	where      string
+	persistent bool
+}
+
+func (mi *ffsMountInfo) validate() error {
+	// for ffs the name is the name of the function, which is not a path so we
+	// just need to ensure it doesn't contain any AppArmor regex characters.
+	if err := validateNoAppArmorRegexpWithError(`cannot use usb-gadget "name" attribute`, mi.name); err != nil {
+		return err
+	}
+
+	// Reuse the where validation from mount-control, as the semantics are the same
+	if err := validateWhereAttr(mi.where); err != nil {
+		return err
+	}
+
+	// Until we have a clear picture of how this should work, disallow creating
+	// persistent mounts into $SNAP_DATA
+	if mi.persistent && strings.HasPrefix(mi.where, "$SNAP_DATA") {
+		return errors.New(`mount-control "persistent" attribute cannot be used to mount onto $SNAP_DATA`)
+	}
+
+	return nil
+}
+
+func ffsMounts(plug interfaces.Attrer) ([]map[string]any, error) {
+	var mounts []map[string]any
+	err := plug.Attr("ffs-mounts", &mounts)
+	if err != nil && !errors.Is(err, snap.AttributeNotFoundError{}) {
+		return nil, mountAttrTypeError
+	}
+	return mounts, nil
+}
+
+func enumerateFFSMounts(mounts []map[string]any, fn func(mi *ffsMountInfo) error) error {
+	for _, mount := range mounts {
+		name, ok := mount["name"].(string)
+		if !ok {
+			return fmt.Errorf(`usb-gadget FunctionFS mount "name" must be a string`)
+		}
+
+		where, ok := mount["where"].(string)
+		if !ok {
+			return fmt.Errorf(`usb-gadget FunctionFS mount "where" must be a string`)
+		}
+
+		persistent := false
+		persistentValue, ok := mount["persistent"]
+		if ok {
+			if persistent, ok = persistentValue.(bool); !ok {
+				return fmt.Errorf(`usb-gadget FunctionFS mount "persistent" must be a boolean`)
+			}
+		}
+
+		mountInfo := &ffsMountInfo{
+			name:       name,
+			where:      where,
+			persistent: persistent,
+		}
+
+		if err := fn(mountInfo); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (iface *usbGadgetInterface) BeforeConnectPlug(plug *interfaces.ConnectedPlug) error {
+	// The systemd.ListMountUnits() method works by issuing the command
+	// "systemctl show *.mount", but globbing was only added in systemd v209.
+	if err := systemd.EnsureAtLeast(209); err != nil {
+		return err
+	}
+
+	mounts, err := ffsMounts(plug)
+	if err != nil {
+		return err
+	}
+	return enumerateFFSMounts(mounts, func(mi *ffsMountInfo) error {
+		return mi.validate()
+	})
+}
+
+func (iface *usbGadgetInterface) AppArmorConnectedPlug(spec *apparmor.Specification, plug *interfaces.ConnectedPlug, slot *interfaces.ConnectedSlot) error {
+	usbGadgetSnippet := bytes.NewBuffer(nil)
+	emit := func(f string, args ...any) {
+		fmt.Fprintf(usbGadgetSnippet, f, args...)
+	}
+	snapInfo := plug.Snap()
+
+	// emit rules required generally for the usb-gadget interface
+	emit(`
+  # https://www.kernel.org/doc/Documentation/usb/gadget_configfs.txt
+  # Allow creating new gadgets under usb_gadget, which is creating
+  # new directories
+  /sys/kernel/config/usb_gadget/ rw,
```

Comment:
This rule only governs the `usb_gadget` directory. I think that here you only need `r` as the directory already exists, right? See similar pattern in snap-confine:

```
    /sys/fs/cgroup/ r,
    /sys/fs/cgroup/devices/ r,
    /sys/fs/cgroup/devices/snap.*/ rw,
    /sys/fs/cgroup/devices/snap.*/cgroup.procs w,
    /sys/fs/cgroup/devices/snap.*/devices.{allow,deny} w,
```

Reply:
Right - but I need to allow the snap to create directories inside that usb_gadget/ -- is this not  the correct way of dealing with that?

Reply:
equivalent to ` /sys/fs/cgroup/devices/snap.*/ rw,` unless I'm missing something?

Reply:
You need permissions on the actual directory to create, not on the parent.